File: triatlon-cs/Model/Activity/Race.cs

// Race

using System.Collections.Generic;
using System.Linq;

public struct Race
{
    
    // Properties

    private RaceType _type;
    private List<AthletePoints> _athletePoints;

    public RaceType Type
    {
        get => _type;
    }

    public List<AthletePoints> AthletePoints
    {
        get => _athletePoints;
    }
    
    // Lifecycle

    public Race(RaceType type)
    {
        _type = type;
        _athletePoints = new List<AthletePoints>();
    }
    
    // Private Methods

    private AthletePoints GetAthletePoints(int athleteId)
    {
        return _athletePoints.FirstOrDefault(e => e.AthleteId == athleteId);
    }

    private bool AthletesPointsContainsAthlete(int athleteId)
    {
        return GetAthletePoints(athleteId) != null;
    }
    
    // Instance Methods

    public void AddAthlete(int athleteId)
    {
        if (AthletesPointsContainsAthlete(athleteId))
        {
            return;
        }
        _athletePoints.Add(new AthletePoints(athleteId));
    }

    public void AddAthlete(int athleteId, int points)
    {
        if (AthletesPointsContainsAthlete(athleteId))
        {
            SetPointsForAthlete(athleteId, points);
            return;
        }
        _athletePoints.Add(new AthletePoints(athleteId, points));
    }

    public void SetPointsForAthlete(int athleteId, int points)
    {
        AthletePoints athletePoints = GetAthletePoints(athleteId);
        if (athletePoints != null)
        {
            athletePoints.Points = points;
        }
    }

    public int? GetPointsForAthlete(int athleteId)
    {
        AthletePoints athletePoints = GetAthletePoints(athleteId);
        return athletePoints?.Points ?? null;
    }
    
}

File: triatlon-java/src/main/java/ro/ubbcluj/scs/dfar2166/triatlonjava/model/activity/Race.java
package ro.ubbcluj.scs.dfar2166.triatlonjava.model.activity;

import ro.ubbcluj.scs.dfar2166.triatlonjava.model.Entity;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

// Race

public class Race extends Entity<Integer> {

    // Private Properties

    private final RaceType type;
    private final List<AthletePoints> athletesPoints;

    // Lifecycle

    public Race(RaceType type) {
        this.type = type;
        this.athletesPoints = new ArrayList<>();
    }

    // Private Methods

    private Optional<AthletePoints> getAthletePoints(Integer athleteId) {
        return athletesPoints.stream().filter(athletePoints -> Objects.equals(athletePoints.getAthleteId(), athleteId)).findFirst();
    }

    private Boolean athletesPointsContainsAthlete(Integer athleteId) {
        return getAthletePoints(athleteId).isPresent();
    }

    // Instance Methods

    public void addAthlete(Integer athleteId) {
        if (athletesPointsContainsAthlete(athleteId)) { return; }
        this.athletesPoints.add(new AthletePoints(athleteId));
    }

    public void addAthlete(Integer athleteId, Integer points) {
        if (athletesPointsContainsAthlete(athleteId)) {
            setPointsForAthlete(athleteId, points);
            return;
        }
        this.athletesPoints.add(new AthletePoints(athleteId, points));
    }

    public void setPointsForAthlete(Integer athleteId, Integer points) {
        Optional<AthletePoints> entity = getAthletePoints(athleteId);
        entity.ifPresent(athletePoints -> athletePoints.setPoints(points));
    }

    public Optional<Integer> getPointsForAthlete(Integer athleteId) {
        return Optional.ofNullable(getAthletePoints(athleteId).get().getPoints());
    }

    // Getters & Setters

    public RaceType getType() {
        return type;
    }

    public List<AthletePoints> getAthletesPoints() {
        return this.athletesPoints;
    }

}
